utils/waitable_counter.hh

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2021-present ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <map>
+#include <iosfwd>
+#include <exception>
+
+#include <seastar/core/future.hh>
+#include <seastar/core/future-util.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/shared_future.hh>
+
+namespace utils {
+
+/**
+ * Simple counter type wrapper, that allows
+ * waiting for it to reach a given value.
+ * 
+ * Actual content must be monotonically 
+ * increasing, and unless you want a lot
+ * of entries, waiting should be either
+ * sparse, or at least quantized to "typical"
+ * values (like disk blocks or similar, hint hint)
+ */
+template<typename T, typename Cmp = std::less<T>>
+class waitable_counter {
+private:
```

Comment:
Missing: unit test

I've a feeling waiting for a condition variable would be more flexible and more efficient (but more effort for the caller).

Reply:
Using a cond or whatnot would result potentially in more spurious wakeups. The idea here again is to only wake up waiters at water marks below the wakers produced position. 
Can be argued how much this is actually relevant of course, need to measure more to be certain.

Reply:
With the fancy coroutine condition variable support, is this still true?

Reply:
I guess it is not. Though technically, every wakeup gets more expensive for each pred needing check, but storage at least is wholly on waiters end. Can change the comment.